enterprise/security/src/test/java/org/neo4j/server/security/enterprise/auth/ShiroCaffeineCacheTest.java

```diff
@@ -0,0 +1,174 @@
+            fail("Expected IllegalArgumentException for a TTL of 0");
+        }
+        catch ( IllegalArgumentException e )
+        {
+            assertThat( e.getMessage(), containsString( "TTL must be larger than zero." ) );
+        }
+        catch ( Throwable t )
+        {
+            fail("Expected IllegalArgumentException for a TTL of 0");
+        }
+    }
+
+    @Test
+    public void shouldNotGetNonExistentValue()
+    {
+        assertThat( cache.get( 1 ), equalTo( null ) );
+    }
+
+    @Test
+    public void shouldPutAndGet()
+    {
+        cache.put( 1, "1" );
+        assertThat( cache.get( 1 ), equalTo( "1" ) );
+    }
+
+    @Test
+    public void shouldNotReturnExpiredValueThroughPut()
+    {
+        assertNull( cache.put( 1, "first" ));
+        assertThat( cache.put( 1, "second" ), equalTo( "first" ) );
+        fakeTicker.advance( TTL + 1, MILLISECONDS );
+        assertNull( cache.put( 1, "third" ) );
+    }
+
+    @Test
+    public void shouldRemove()
+    {
+        assertNull( cache.remove( 1 ) );
+        cache.put( 1, "1" );
+        assertThat( cache.remove( 1 ), equalTo( "1" ) );
+    }
+
+    @Test
+    public void shouldClear()
+    {
+        cache.put( 1, "1" );
+        cache.put( 2, "2" );
+        assertThat( cache.size(), equalTo( 2 ) );
+        cache.clear();
+        assertThat( cache.size(), equalTo( 0 ) );
+    }
+
+    @Test
+    public void shouldGetKeys()
+    {
+        cache.put( 1, "1" );
+        cache.put( 2, "1" );
+        cache.put( 3, "1" );
+        assertThat( cache.keys(), containsInAnyOrder( 1, 2, 3 ) );
+    }
+
+    @Test
+    public void shouldGetValues()
+    {
+        cache.put( 1, "1" );
+        cache.put( 2, "1" );
+        cache.put( 3, "1" );
+        assertThat( cache.values(), containsInAnyOrder( "1", "1", "1" ) );
+    }
+
+    @Test
+    public void shouldNotListExpiredValues()
+    {
+        cache.put( 1, "1" );
+        fakeTicker.advance( TTL + 1, MILLISECONDS );
+        cache.put( 2, "foo" );
+
+        assertThat( cache.values(), containsInAnyOrder( "foo" ) );
+    }
+
+    @Test
+    public void shouldNotGetExpiredValues()
+    {
+        cache.put( 1, "1" );
+        fakeTicker.advance( TTL + 1, MILLISECONDS );
+        cache.put( 2, "foo" );
+
+        assertThat( cache.get( 1 ), equalTo( null ) );
+        assertThat( cache.get( 2 ), equalTo( "foo" ) );
+    }
+
+    @Test
+    public void shouldNotGetKeysForExpiredValues()
+    {
+        cache.put( 1, "1" );
+        fakeTicker.advance( TTL + 1, MILLISECONDS );
+        cache.put( 2, "foo" );
+
+        assertThat( cache.keys(), containsInAnyOrder( 2 ) );
+    }
+
+    @Test
+    public void shouldRemoveIfExceededCapacity()
+    {
+        cache.put( 1, "one" );
+        cache.put( 2, "two" );
+        cache.put( 3, "three" );
+        cache.put( 4, "four" );
+        cache.put( 5, "five" );
+        cache.put( 6, "six" );
+
+        assertThat( cache.keys(), containsInAnyOrder( 2, 3, 4, 5, 6 ) );
```

Comment:
Because the algorithm isn't LRU and has some randomness to protect against HashDoS attacks, this could be flaky. The random hash seed could cause a collision in the frequency sketch resulting in `1` being retained. You might prefer to check the size instead of the contents.